client/dashboard/task-list/tasks/tax.js

```diff
@@ -148,10 +149,10 @@ class Tax extends Component {
 	}
 
 	async updateAutomatedTax() {
-		const { createNotice, isTaxSettingsError, updateSettings } = this.props;
+		const { createNotice, isTaxSettingsError, updateAndPersistSettingsForGroup } = this.props;
 		const { automatedTaxEnabled } = this.state;
 
-		await updateSettings( {
+		await updateAndPersistSettingsForGroup( 'tax', {
```

Comment:
@nerrad, here I have a call to update and persist a settings group, which works fine in other instances. The difference in this one is that the `tax` settings group is instantiated via an API call instead of hydration via `wcSettings`, which is all working as it should.

This persistence call creates an error on WP 5.3.2, but not on the latest version of Gutenberg.

![Screen Shot 2020-02-12 at 5 02 54 PM](https://user-images.githubusercontent.com/1922453/74302093-146a7b80-4dba-11ea-9595-c7f9e3598551.png)

I traced the fix back to this pull request: https://github.com/WordPress/gutenberg/pull/17558

As someone who reviewed that PR, would you happen to have any insight as to why this error is happening in the previous to changes in that PR? I'm curious because there are backward compat issues, unless this is pointing to something I'm doing wrong.

Reply:
I'm not sure how the linke pull relates to this issue here because it is dealing with a _selector_ not an action which is what is causing the error for you (if indeed line 155 here is the cause). So that might be a red herring.  However, the one thing it _might_ be related to is that `updateAndPersistSettingsForGroup` internally results in the select control being called which _is_ modified by the linked pull you pointed to.  Essentially it _looks_ from the error like a requested selector/action does not exist in the store at the time redux/routine is trying to call it. You could review all selector, action and store names to make sure that they don't have any typos?

Another potential issue here is that ultimately you cannot rely on the response from a dispatched action (which you are doing with await) - across various versions of WordPress core and Gutenberg, wp.data dispatched actions return different final values and I'm pretty sure the GB team is moving towards always returning undefined. So to rely on the return value is fragile.

Instead the action should be dispatched directly and then your component reacts however it needs once the state is updated by the dispatched action. So that may mean what state gets updated by the dispatched action, it may mean you need to refactor the component to account for that, it may mean introducing a new action/selector and using that. I'm not sure exactly what might be needed in this case without doing a deep dive (which I'm willing to do if needed).

Another potential problem here is that I see `setSetting` is being called for the `onboarding` object.  Doesn't that mutate the global? I wonder if that's causing something persisted in the data store to get mutated which results in this intermediate value error (unlikely, but probably still should be handled...I'm guessing this is just something not refactored yet)?

Reply:
Thanks for taking a look @nerrad .

>However, the one thing it might be related to is that `updateAndPersistSettingsForGroup` internally results in the select control being called which is modified by the linked pull you pointed to.

Right, so that action calls `persistSettingsForGroup` which has two selects (in pseudo code):

```js
import { select } from '@wordpress/data-controls';

const dirtyKeys = yield select( STORE_NAME, 'getDirtyKeys', group );
const dirtyData = yield select( STORE_NAME, 'getSettingsForGroup', group, dirtyKeys );
```
The first works fine, but the second does not. Here is part of the faulty line:

```js
registry[registry.select(storeKey)[selectorName].hasResolver ? '__experimentalResolveSelect' : 'select']
```
The second selector `getSettingsForGroup` has a resolver, so the ternary expression results in `__experimentalResolveSelect`.

Stepping through the code, `registry.__experimentalResolveSelect` does not exist in the old version of Gutenberg, but does after the merge of https://github.com/WordPress/gutenberg/pull/17558. 

Here is the problematic function (from a commit before the above PR was merged):
https://github.com/WordPress/gutenberg/blob/4857ad58c1241b3d63d21a6880c989b85746c3dc/packages/data-controls/src/index.js#L127-L135

Why does the selector need to use `__experimentalResolveSelect`? That selector has already resolved and its results already in state. I can't gather from the PR why it is defined in newer versions of Gutenberg, but not in the one shipped with 5.3.2.

>Another potential issue here is that ultimately you cannot rely on the response from a dispatched action (which you are doing with await) - across various versions of WordPress core and Gutenberg, wp.data dispatched actions return different final values and I'm pretty sure the GB team is moving towards always returning undefined. So to rely on the return value is fragile.

Good point. I think thats unrelated to these changes, but can be addressed in a follow up.

>Another potential problem here is that I see setSetting is being called for the onboarding object. Doesn't that mutate the global?

Yes, thats a problem. But my approach is to handle the mutable settings of `wcSettings` one at a time. This pull addresses `preloadSettings`. Up next will be `onboarding` and `currentUserData`. This way I can remove `wc-api` one folder at a time and thus have incremental pull requests from master after this series of PRs is merged.

Reply:
If I copy the problematic selector function `getSettingsForGroup` and rename it `getAlreaydResolvedSettingsForGroup` such that it does not have a matching resolver by the same name, everything works as expected.

Although this could be a work-around, it really seems dirty and an indication that something is wrong.